### PR TITLE
Expose marker data

### PR DESCRIPTION
**Pull request by paulsonnentag**
This pull request changes two things:

- Adds `paperNumber` property to markers
- Paper Programs detects the markers by size and not by the fact that they are not part of a shape. When the user calibrates the colors of the markers, Paper Programs stores the size of the markers as well. Dots which are significantly larger than the paper dots are recognized as markers. It's no longer necessary to filter the markers for size in the paper program to get the real markers.

I haven't implemented the local position conversion, because I realized that it's possible that a paper has multiple canvases with different sizes. I think it's better to have a utility function to convert global coordinates to local coordinates. I want to wait until #44 is merged which restructures `paper.js` to allow imports.

### CODE REVIEW

**Comment by shaunlebron** [comment]
> I realized that it's possible that a paper has multiple canvases with different sizes. I think it's better to have a utility function to convert global coordinates to local coordinates

I was wanting something like this too.  Might make sense to attach a projection matrix to each canvas returned by `paper.get('canvas', ...)`

File: client/utils.js [hunk 1]
```diff
@@ -34,6 +34,10 @@ export function clamp(value, min, max) {
   return Math.max(min, Math.min(max, value));
 }
 
+export function sign(value) {
+  return value < 0 ? -1 : value > 0 ? 1 : 0;
```
Comment (shaunlebron) [inline client/utils.js]: we can use [Math.sign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/sign)

Comment (paulsonnentag) [reply client/utils.js]: Ah thanks, didn't know about that

File: client/camera/detectPrograms.js [hunk 2]
```diff
@@ -404,26 +412,49 @@ export default function detectPrograms({ config, videoCapture, dataToRemember, d
...
+      for (let i = 0; i < 4; i++) {
+        const a = i;
+        const b = (i + 1) % 4;
+
+        const sideA = diff(points[a], markerPosition);
+        const sideB = diff(points[b], markerPosition);
+
+        let angle = Math.atan2(sideB.y, sideB.x) - Math.atan2(sideA.y, sideA.x);
+
+        if (sign(sideB.y) === -1 && sign(sideA.y) === 1) {
```
Comment (janpaul123) [inline client/camera/detectPrograms.js]: `if(sideB.y < 0 && sideA.y > 0)`?

**Comment by paulsonnentag** [comment]
I need some help with transforming the point to local coordinates. I don't really understand how the matrix calculations work. I thought I could do it like this:

```JavaScript
const projectionMatrix = forwardProjectionMatrixForPoints(program.points)
const relativePosition = projectPoint(marker.position, projectionMatrix)
```

Both marker.position and program.points are the original coordinates from detectProgram (value range between 0 and 1). I expected to get a transformed position relative to the top left of the paper (value range between 0 and 1) but this doesn't work.

**Comment by shaunlebron** [comment]
I learned about it mainly from the figure below in this [paper], and experimented with it in this [notebook]:

screen shot 2018-05-03 at 9 06 06 pm

- case 1 is a forward projection
- case 2 is an inverse projection
- case 3 is an arbitrary projection created by multiplying matrix 1 and 2 (neat!)

The forward projection (case 1 above) maps _unit-square_ coords to those inside the given quad corners.  For example, it maps `1,1` to corner 2.

In context, if you take a local coord inside the paper's canvas, divide it by the canvas size, then project it with the result of `forwardProjectionMatrixForPoints` you will get its coordinates on the supporter canvas.

This marker problem requires the _inverse_ (case 2 above).  Luckily, the [paper] describes that the inverse of a projection mapping is itself a projection mapping, and it can be created by taking the adjugate:

```js
const forward = forwardProjectionMatrixForPoints(program.points);
const inverse = forward.adjugate();

let relativePosition = projectPoint(marker.position, inverse);
relativePosition = mult(relativePosition, {x: canvas.width, y: canvas.height}); // rescale
```

And actually, you can avoid that rescaling part by multiplying it with a "canvas size" matrix, though I can never remember the right order (sorta like case 3 above).  Example is here though:

https://github.com/janpaul123/paperprograms/blob/c4eeb5a57ff0002f40541c87da2d623e59bab5e0/client/projector/Program.js#L179-L181

[paper]:http://graphics.cs.cmu.edu/courses/15-463/2008_fall/Papers/proj.pdf
[notebook]:https://beta.observablehq.com/@shaunlebron/paper-programs-matrix-projection

**Comment by janpaul123** [comment]
@paulsonnentag would you like me to merge this PR and then follow up with local coordinates in a second PR (after you figure out the matrix stuff) or want to do it all at once?

**Comment by paulsonnentag** [comment]
I would like to finish the local coordinates first and also update the documentation

**Comment by paulsonnentag** [comment]
@janpaul123 ready to merge

**Comment by janpaul123** [comment]
So cool. Can't wait to try this out soon!

**Pull request merged**